File: Python/insercion_binaria.py
# Programa para ordenar una lista mediante insertion sort y buscar un elemento por binary search

# Funcion para llenar una lista

def fill_list(lista):
    length = int(input("\nIngrese la cantidad de elementos de la lista: "))

    for i in range(0, length):
        value = int(input(f"Ingrese el elemento numero {i}: "))
        lista.append(value)

# Funcion para el insertion sort

def insertion_sort(lista, size):

    for i in range(1, size):
        t = lista[i]
        j = i
        while j > 0 and t < lista[j - 1]:
            lista[j] = lista[j - 1]
            j -= 1

        lista[j] = t

# Funcion de la binary search

def binary_search(lista, size, v):

    start = 0
    end = size - 1

    while start <= end:
        mid = (start + end) // 2

        if lista[mid] == v:
            return mid
        elif lista[mid] < v:
            start = mid + 1
        else:
            end = mid - 1

    return -1

# Funcion main()

int_list = []

# Llamada a funcion para llenar la lista

fill_list(int_list)
print(f"\nLa lista es: {int_list}")
list_size = len(int_list)

# Llamada a funcion para el insertion sort

print("\nOrdenando por insertion sort...\n")
insertion_sort(int_list, list_size)
print(f"La lista ordenada es: {int_list}\n")

# Llamada a funcion para la binary search

value = int(input("Ingrese el valor del elemento a buscar: "))
pos = binary_search(int_list, list_size, value)
print(f"\nLa posicion del elemento {value} es {pos}\n")

File: C/insercion_binaria.c
// Programa para ordenar un arreglo mediante insertion sort y buscar un elemento mediante binary_search

#include <stdio.h>

// Funcion para llenar un arreglo

void fill_array(int arr[], int size)
{
    register int i;
    int element;

    for (i = 0; i < size; i++)
    {
        printf("\nIngrese el valor del elemento %d ", i);
        scanf("%d", &element);
        arr[i] = element;
    }
    printf("\n");
}

// Funcion para imprimir un arreglo

void print_array(int arr[], int size)
{
    register int i;

    printf("\nLos elementos del arreglo son: [ ");
    
    for (i = 0; i < size; i++)
    {
        printf("%d ",arr[i]);
    }
    
    printf("]\n");
}

void insertion_sort(int arr[], int size)
{
    register int i, j;
    int t;

    for (i = 1; i < size; i++)
    {
        t = arr[i];
        for (j = i; j > 0 && t < arr[j - 1]; j--)
            arr[j] = arr[j - 1];
        
        arr[j] = t;
    }
}

// Funcion para la binary search

int binary_search(int arr[], int size, int v)
{
    int start, end, mid;

    start = 0;
    end = size - 1;

    while (start <= end)
    {
        mid = (start + end) / 2;

        if (arr[mid] == v)
            return mid;
        else if (arr[mid] < v)
            start = mid + 1;
        else
            end = mid - 1;
    }
}

void main()
{
    int n, value, pos;

    printf("\nIngrese la cantidad de elementos del arreglo: ");
    scanf("%d", &n);

    int array[n];

    // Llamada a funcion para llenar los elementos del arreglo

    fill_array(array, n);
    print_array(array, n);
    
    // Llamada a funcion para ordenar por insertion sort

    printf("\nOrdenando por insertion sort...");
    insertion_sort(array, n);
    print_array(array, n);

    // Llamada a funcion para buscar un elemento mediante binary search

    printf("\nIngrese el valor del elemento a buscar: ");
    scanf("%d", &value);

    pos = binary_search(array, n, value);
    printf("\nLa posicion del elemento %d es: %d", value, pos);
}